### DISCUSSION TITLE
How to convert PNG/SVG to XBM (Xml bitmap)

### DISCUSSION THREAD

**Discussion by ShivamJoker** [ImageMagick/ImageMagick#4138] [Help] [answered]
I tried running the convert command but it just generates the 0000

![image](https://user-images.githubusercontent.com/23727670/131551346-9e9fb70f-6258-4ff1-a092-12831ec0323d.png)

Here is the png (it's very small)

![icons8-empty-battery-24](https://user-images.githubusercontent.com/23727670/131551439-9c8328eb-87ec-4649-9c80-806a7d455117.png)

Do I need to do anything else ?

**Comment by snibgo** [comment]
Your input is entirely black, with varying opacity. XBM seems to store each pixel as a single bit, so ignores opacity.

**Reply by ShivamJoker** [reply]
But how do I fix it ? I mean do I need to convert it into something else ?

**Comment by snibgo** [comment]
XBM seems to be a simple format that can record black and white only, with no transparency.

Perhaps you really want XPM format, which can record colours and transparency.

**Reply by ShivamJoker** [reply]
okay so you mean i need to convert it into jpeg and try again ?

**Comment by snibgo** [comment]
Huh? No.

Why do you want XBM format? It doesn't do what you want. But XPM format does do what you want.
```
convert icons8-empty-battery-24.png out.xpm
```

out.xpm is a text file:
```
/* XPM */
static char *out[] = {
/* columns rows colors chars-per-pixel */
"24 24 2 1 ",
"  c black",
". c None",
/* pixels */
"........................",
"........................",
".........      .........",
".........      .........",
".......    ..    .......",
"......     ..     ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......  ........  ......",
"......            ......",
".......          .......",
"........................",
"........................"
};
```

**Comment by snibgo** [comment]
By the way, your title is "How to convert PNG/SVG to XBM (Xml bitmap)". But that is wrong. The X is for "X Window", not "XML". The text format is C code, not XML code.

**Reply by ShivamJoker** [reply]
You got my question, right ? 😐
It would be really helpful if you could simply answer what I need

**Comment by snibgo** [comment, marked as answer]
Sorry, I thought I had answered your question. Your image contains only opaque black and transparent black, no white. XBM can record only black and white, and does not record transparency. So the result XBM file is, correctly, all black.

If you want a black symbol on a white background, you can flatten the image against a white background before writing the XBM file:
```
convert icons8-empty-battery-24.png -background White -layers Flatten out.xbm
```
out.xbm is then:
```
#define out_width 24
#define out_height 24
static char out_bits[] = {
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x7E, 0x00, 0x00, 0x7E, 0x00,
  0x80, 0xE7, 0x01, 0xC0, 0xE7, 0x03, 0xC0, 0x00, 0x03, 0xC0, 0x00, 0x03,
  0xC0, 0x00, 0x03, 0xC0, 0x00, 0x03, 0xC0, 0x00, 0x03, 0xC0, 0x00, 0x03,
  0xC0, 0x00, 0x03, 0xC0, 0x00, 0x03, 0xC0, 0x00, 0x03, 0xC0, 0x00, 0x03,
  0xC0, 0x00, 0x03, 0xC0, 0x00, 0x03, 0xC0, 0x00, 0x03, 0xC0, 0x00, 0x03,
  0xC0, 0xFF, 0x03, 0x80, 0xFF, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  };
```

**Reply by ShivamJoker** [reply]
Thanks this is what I wanted.
